### PR TITLE
Introduce `JUnitSingleArguments` bug checker

### PR DESCRIPTION
**Pull request by Venorcis**
Adds a bugchecker that flags uses of JUnit `Argument`s when only a single (or in a weird edge case no) argument is used.

An example can be seen in the diff of the `MethodMatcherFactoryTest` 😄

### CODE REVIEW

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.jupiter.api.Test;
+
+final class JUnitSingleArgumentsTest {
+  @Test
+  void identification() {
+    CompilationTestHelper.newInstance(JUnitSingleArguments.class, getClass())
+        .addSourceLines(
```
Comment (Venorcis) [inline error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java:10]: I don't get what's not formatting-compliant here 🤔

Comment (Stephan202) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java:10]: It was subtle; ran `./apply-error-prone-suggestions.sh` and pushed :)

Comment (Venorcis) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java:10]: Ah thanks, didn't know that existed 😄 Also the command line error just gives 'did you mean' with the first few lines, which is confusing 🤪

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  private static final Matcher<ExpressionTree> ARGUMENTS_ARGUMENTS =
+      staticMethod().onClass("org.junit.jupiter.params.provider.Arguments").named("arguments");
+
+  /** Instantiates a new {@link JUnitSingleArguments} instance. */
+  public JUnitSingleArguments() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (ARGUMENTS_ARGUMENTS.matches(tree, state) && tree.getArguments().size() <= 1) {
+      return describeMatch(tree, SourceCode.unwrapMethodInvocation(tree, state));
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: This would cause a compilation error w.r.t. the return type, right? :thinking:

**Review by Stephan202 — COMMENTED** [review]
Nice rule! I suspect we an make it robust in the common case, but will need to find some time to look closer into the least-impactful approach.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java (hunk 2 above)
Comment (Venorcis) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: Yeah, I wasn't sure how we deal with the fix suggestions. AFAIK not all suggestions necessarily lead to directly compiling code? We can ofc. remove it here and just flag instead 🤷🏻‍♂️

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java (hunk 1 above)
Comment (Stephan202) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArgumentsTest.java:10]: Yeah, that output is... "not great". Haven't really found a way yet to make that play nicer with Error Prone.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java (hunk 2 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: We _try_ either produce compilable code _or_ only flag the issue, with a strong preference for the former (as the latter really introduces quite a bit more friction). That said, especially some of the Refaster rules can in special circumstances break the code.

For this case we could:
1. Simplify the factory method if it has a single return statement, using similar logic we use in [`JUnitValueSource`](https://github.com/PicnicSupermarket/error-prone-support/blob/da5eea832975cb9222cba51935f7a22b4ff35e46/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitValueSource.java).
2. Only flag otherwise.

(But that's considerably more complicated than the current proposal, so if you're like "uuuuuh", I get that. Don't mind having a stab; would just take a bit of time.)

Comment (Venorcis) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: Yeah understood. In the most common use case we 'only' need to replace `<Arguments>` with `<T>` in the method definition. Perhaps it's relatively easy to add the suggestion for this use case and still resort to flagging any other occurrences (they would be weird but you never know 😛 ).

Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: Indeed, that's the lines along which I'm thinking :+1:

(Both inside Picnic and for the upcoming integration test framework against open-source repos we run Error Prone in a loop until it no longer makes any changes, which can only work if each intermediate state compiles successfully.)

Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitSingleArguments.java:45]: I've looked a bit closer at this topic, and the logic in `JUnitValueSource` shows that there are quite a lot of details to consider:
1. There are different ways in which the `Arguments` can be wrapped in a top-level collection/stream/array.
2. The factory method can have multiple return statements.
3. The return type could change from `MonadType<Arguments>` or `Arguments[]` to `MonadType<Object>` or `Object[]`, but ideally we're more specific.
4. Only if all `Arguments` instances have exactly one value can we do the unwrapping (in case of variability we should likely emit a separate warning).
5. Ideally we do all this in a way that reuses relevant pieces of logic from `JUnitValueSource`.

I'm likely missing a few other details. Happy to dive deeper into this, but it'll be a while before I find the time to really sit down for this.
